GoogleForCreators/web-stories-wp
# Dashboard: Typeahead UI Component

# Overview
Adds a typeahead component for use in Dashboard for filtering page results. Returns value of selection or input on change, expects available items for selection to be filtered by parent either from the server at once or real time - I have included a helper function to handle filtering if all data is received at once and we need to handle for that on the front end. 

## Changes 
- Creates `TypeaheadInput` component, works a lot like the `Dropdown` but is styled very differently. 
  - SearchContainer holds an input for user to type into and a `TypeaheadOptions` (looks alot like `PopoverMenu` but again, styled differently enough to warrant separation). 
  - Button to clear input only shows up when there's a value in the input 
  - Menu only shows up once there's value in input and there's items available to show in the menu
- Creates `TypeaheadOptions` - the dropdown that shows up when input has value and is focused
  - Functions a lot like `PopoverMenu` but not entirely, this one has the ability to limit results 
- swaps black fills for currentColor on close and search icons 
- Adds boxShadow for typeahead to theme, I put this here to identify patterns in the boxShadow early on so we can reuse and gain continuity. 
- Adds l[odash.memoize](https://lodash.com/docs/4.17.15#memoize) to package for `getTypeaheadOptions` ([file](https://github.com/google/web-stories-wp/pull/750/files#diff-b281bc916678c07be6962a04352555a2)) - if we don't need to filter on the front end at all then we can get rid of this. 

## Testing 
- run npm install so you have lodash.memoize available and run storybook locally 
- see [Dashboard/Components/TypeaheadInput](http://localhost:53006/?path=/story/dashboard-components-typeaheadinput--default) and [DashboardComponents/TypeaheadOptions](http://localhost:53006/?path=/story/dashboard-components-typeaheadoptions--default)
- TypeaheadInput: Default
  - Expects items shown when menu is visible to be filtered in real time, so they aren't going to ever update with what's getting typed. 
- TypeaheadInput: Filter Loaded Search Items 
  - Should filter items visible in menu based on what you type into the input. Is checking both label and value of menu items, so there's one item in the demo that has { value: 'agua', label: 'water' } - both of those will give you the same 'water' in the input value to show.  If you continue typing "waterrrrr" then that is going to be the value sent to the parent. Since we don't know how exactly the data will run through this, I figure if this is an issue we can circle back. I wanted to allow flexible typing in there for now. 
Both demos: 
  - Should be able to select an item with mouse and with keyboard 
  - Should be able to type whatever you want in there and when you exit the input it should stay 
  - Should clear input value when 'x' is clicked
  - 'x' should only be visible when input has value 

- TypeaheadOptions
  - This is just the menu shown when the TypeaheadInput is open. 
  - Clicking on an item will log an action in the actions dropdown 

**Default**
![uncontrolled](https://user-images.githubusercontent.com/10720454/77474554-9051dd80-6dd4-11ea-92b7-8ab9472c7dee.gif)

**With filtering**
![filtering](https://user-images.githubusercontent.com/10720454/77474521-816b2b00-6dd4-11ea-983a-44a252b20d7f.gif)

**Expanded Typeahead** 
Code: 
![Screen Shot 2020-03-24 at 1 42 44 PM](https://user-images.githubusercontent.com/10720454/77475077-74027080-6dd5-11ea-948d-3924584cf73c.png)

Figma:
![Screen Shot 2020-03-24 at 1 39 43 PM](https://user-images.githubusercontent.com/10720454/77474794-f76f9200-6dd4-11ea-9fc5-4ffd34c2ec29.png)

**Collapsed Typeahead**
Code:
![Screen Shot 2020-03-24 at 1 39 01 PM](https://user-images.githubusercontent.com/10720454/77474744-e0c93b00-6dd4-11ea-8e97-cfe330e4a2a3.png)

Figma:
![Screen Shot 2020-03-24 at 1 39 48 PM](https://user-images.githubusercontent.com/10720454/77474811-fd657300-6dd4-11ea-9791-1ead7cd34196.png)


## Todo 
- I'm not happy with the outline for keyboards on this input but I want to circle back to it since it's not critical to our initial release.
